query/influxql/result.go

```
@@ -28,106 +26,103 @@ func (e *MultiResultEncoder) Encode(w io.Writer, results query.ResultIterator) e
-							}
-							r.Name = cr.Strings(j)[i]
-							continue
-						}
+			if err := b.Do(func(cr query.ColReader) error {
+				var values [][]interface{}
+				i := 0
+				for idx, c := range b.Cols() {
+					if cr.Key().HasCol(c.Label) {
+						continue
+					}
 
+					// Use the first row, usually time, to pre-generate all of the values.
+					if i == 0 {
 						switch c.Type {
-						case query.TFloat:
-							v = append(v, cr.Floats(j)[i])
-						case query.TInt:
-							v = append(v, cr.Ints(j)[i])
-						case query.TString:
-							v = append(v, cr.Strings(j)[i])
-						case query.TUInt:
-							v = append(v, cr.UInts(j)[i])
-						case query.TBool:
-							v = append(v, cr.Bools(j)[i])
 						case query.TTime:
-							v = append(v, cr.Times(j)[i].Time().Format(time.RFC3339))
+							values = make([][]interface{}, len(cr.Times(0)))
 						default:
-							v = append(v, "unknown")
+							// TODO(jsternberg): Support using other columns. This will
+							// mostly be necessary for meta queries.
+							return errors.New("first column must be time")
+						}
+
+						for i := range values {
+							values[i] = make([]interface{}, len(r.Columns))
 						}
 					}
 
-					r.Values = append(r.Values, v)
+					// Fill in the values for each column.
+					switch c.Type {
+					case query.TFloat:
+						for j, v := range cr.Floats(idx) {
+							values[j][i] = v
+						}
+					case query.TInt:
+						for j, v := range cr.Ints(idx) {
+							values[j][i] = v
+						}
+					case query.TString:
+						for j, v := range cr.Strings(idx) {
+							values[j][i] = v
+						}
+					case query.TUInt:
+						for j, v := range cr.UInts(idx) {
+							values[j][i] = v
+						}
+					case query.TBool:
+						for j, v := range cr.Bools(idx) {
+							values[j][i] = v
+						}
+					case query.TTime:
+						for j, v := range cr.Times(idx) {
+							values[j][i] = v.Time().Format(time.RFC3339)
+						}
+					default:
+						// Unknown type. Leave it as nil.
```

This should return an error. It will be the transpiler's responsibility to ensure that the results only have columns that this encoder supports.